corgidrp/sorting.py

```diff
@@ -0,0 +1,232 @@
+            exptime_list = []
+            gain_filepath_list = []
+            for frame in gain_set:
+                frame_id_list += [extract_frame_id(frame.filename)]
+                exptime_list += [frame.ext_hdr['EXPTIME']]
+                gain_filepath_list += [frame.filepath]
+            # One can set a stronger condition, though in the end the max set
+            if len(frame_id_list) < 3:
+                continue
+            idx_id_sort = np.argsort(frame_id_list)
+            exptime_arr = np.array(exptime_list)[idx_id_sort]
+            # We need an increasing series of exposure times with the last one
+            # the only repeated value in the series
+            idx_subsets = np.where(np.diff(exptime_arr) < 0)[0]
+            if len(idx_subsets) == 0:
+                continue
+            # length of candidate subsets
+            nonlin_len = []
+            for idx, idx_subset in enumerate(idx_subsets):
+                # Add 0th element plus the one lost in diff
+                if idx == 0:
+                    exptime_tmp = exptime_arr[0:idx_subset+2]
+                else:
+                    exptime_tmp = exptime_arr[idx_subsets[idx-1]+1:idx_subset+2]
+                # Check conditions
+                if (exptime_tmp[-1] in exptime_tmp[:-1] and
+                    len(exptime_tmp) - 1 == len(set(exptime_tmp))):
+                    nonlin_len += [len(exptime_tmp)]
+                else:
+                    nonlin_len += [-1]
+            # COntinue of there are no good candidates
+            if np.max(nonlin_len) <= 0:
+                continue
+            # Find maximum set among good candidates
+            if np.argmax(nonlin_len) == 0:
+                idx_nonlin_first = 0
+                idx_nonlin_last = idx_subsets[0] + 1
+            else:
+                idx_nonlin_first = idx_subsets[np.argmax(nonlin_len)-1] + 1
+                idx_nonlin_last = idx_subsets[np.argmax(nonlin_len)] + 1
+            # Sort unity gain filenames
+            gain_filepath_arr = np.array(gain_filepath_list)[idx_id_sort]
+            cal_list = gain_filepath_arr[idx_nonlin_first:idx_nonlin_last+1]
+            # Update OBSTYPE and take profit to check files are in the list
+            for frame in dataset_cp:
+                if frame.filepath in cal_list:
+                    frame.pri_hdr['OBSTYPE'] = 'NONLIN'
+                    cal_frame_list += [frame]
+                    n_nonlin += 1
+            nonlin_emgain += [split_cmdgain[1][idx_gain_set]]
+
+        print(f'Non-linearity has {n_nonlin} frames with gains', nonlin_emgain)
+        
+    # EM-gain: TODO
+    elif cal_type.lower() == 'em-gain' or cal_type.lower() == 'emgain':
+        print('Considering low EM-gain: TODO')
+    else:
+        raise Exception('Unrecognized calibration type (expected k-gain, non-lin or em-gain)')
+
+    # Return Datafrane with mean frame and cal type
+    return data.Dataset(mean_frame_list + cal_frame_list)
```

Comment:
before returnign the dataset, it will be good to add a HISTORY entry to the ext_hdr (like other step functions) that describe what sorting was done.

Reply:
@semaphoreP  I do not know enough about corgidrp. Not sure how to do this. My experience is with step functions that are a derivative of Image and have in its creation the pri_hdr and ext_hdr keywords. I have reviewed the step functions in corgidrp and that's what I've found.

In this case, I'm using Dataset.  Also, each frame of this dataset has a modified primary header with a specific value of OBSTYPE, and EXPTIME, CMDGAIN in its extension header. I am not sure if creating the final Dataset object based on a single ext header is going to change the values of all the frames.

Given the time left to finish several other tasks related with this PR, can you write how it should be?

Reply:
if you have a dataset object, just run `dataset.update_after_processing_step("what was done in this step")`. Doing this will write "what was done in this step" into the history of every image in the dataset.

Reply:
Oh yeah. Long time I had not seen it.

Reply:
Done